src/Storages/MergeTree/KeeperAccess.h

```diff
@@ -0,0 +1,236 @@
+#pragma once
+#include <random>
+#include <Common/ZooKeeper/IKeeper.h>
+#include <Common/ZooKeeper/KeeperException.h>
+#include <Common/ZooKeeper/Types.h>
+#include <Common/ZooKeeper/ZooKeeper.h>
+#include <Common/randomSeed.h>
+
+namespace DB
+{
+
+class FaultInjection
+{
+public:
+    explicit FaultInjection(UInt64 seed_ = 0) : seed(seed_)
+    {
+        if (0 == seed)
+            seed = randomSeed();
+    }
+    virtual ~FaultInjection() = default;
+    virtual void beforeOperation() = 0;
+    virtual void afterOperation() = 0;
+    UInt64 getSeed() const { return seed; }
+
+protected:
+    UInt64 seed = 0;
+};
+
+class Random : public FaultInjection
+{
+public:
+    static std::unique_ptr<Random> create(UInt64 seed_, double probability)
+    {
+        if (probability < 0.0)
+            probability = .0;
+        if (probability > 1.0)
+            probability = 1.0;
+
+        return std::make_unique<Random>(seed_, probability);
+    }
+
+    explicit Random(const UInt64 seed_, double probability) : FaultInjection(seed_), rndgen(getSeed()), distribution(probability) { }
+    ~Random() override = default;
+
+    void beforeOperation() override
+    {
+        if (distribution(rndgen))
+            throw zkutil::KeeperException("Fault injection", Coordination::Error::ZSESSIONEXPIRED);
+    }
+    void afterOperation() override
+    {
+        if (distribution(rndgen))
+            throw zkutil::KeeperException("Fault injection", Coordination::Error::ZOPERATIONTIMEOUT);
+    }
+
+private:
+    std::mt19937_64 rndgen;
+    std::bernoulli_distribution distribution;
+};
+
+class KeeperAccess
+{
+    using zk = zkutil::ZooKeeper;
+
+    zk::Ptr keeper;
+    std::unique_ptr<FaultInjection> fault_policy;
+    std::string name;
+    Poco::Logger * logger = nullptr;
+    UInt64 calls_total = 0;
+    UInt64 calls_without_fault_injection = 0;
+    UInt64 seed = 0;
+
+public:
+    using Ptr = std::shared_ptr<KeeperAccess>;
+
+    static KeeperAccess::Ptr createInstance(
```

Comment:
That is because `make_shared` requires public constructor. But in case of fault injection, constructing object in less efficient way will not hurt, so ok, will address it